How to Select Exists ?

How to write following query ?

```sql
SELECT EXISTS (SELECT 1)
```

One try:
```go
lk.Select(lk.Exists(lk.Select("1"))).String()
```

It fails like this:
```sh
panic: loukoum: cannot use stmt.Exists as column
```

I checked tests and it was only tested as a condition. Missing anything ?

--- a/builder/builder.go
+++ b/builder/builder.go
@@ -116,6 +116,62 @@ func ToColumns(values []interface{}) []stmt.Column { // nolint: gocyclo
 	return columns
 }
 
+// ToSelectExpressions takes a list of empty interfaces and returns a slice of SelectExpression instance.
+func ToSelectExpressions(values []interface{}) []stmt.SelectExpression { // nolint: gocyclo
+	// If values is a slice, we try to use recursion to obtain a slice of Column.
+	if len(values) == 1 {
+		switch array := values[0].(type) {
+		case []stmt.SelectExpression:
+			return array
+		case []stmt.Column:
+			expressions := make([]stmt.SelectExpression, 0, len(values))
+			for i := range array {
+				if array[i].IsEmpty() {
+					panic("loukoum: given column is undefined")
+				}
+				expressions = append(expressions, array[i])
+			}
+			return expressions
+		case []string:
+			expressions := make([]stmt.SelectExpression, 0, len(values))
+			for i := range array {
+				expressions = append(expressions, stmt.NewColumn(array[i]))
+			}
+			return expressions
+		}
+	}
+
+	columns := make([]stmt.SelectExpression, 0, len(values))
+
+	for i := range values {
+		switch value := values[i].(type) {
+		case stmt.SelectExpression:
+			if value.IsEmpty() {
+				panic("loukoum: given column is undefined")
+			}
+			columns = append(columns, value)
+		case stmt.Column:
+			if value.IsEmpty() {
+				panic("loukoum: given column is undefined")
+			}
+			columns = append(columns, value)
+		case string:
+			array := strings.Split(value, ",")
+			for y := range array {
+				column := stmt.NewColumn(strings.TrimSpace(array[y]))
+				if column.IsEmpty() {
+					panic("loukoum: given column is undefined")
+				}
+				columns = append(columns, column)
+			}
+		default:
+			panic(fmt.Sprintf("loukoum: cannot use %T as column", values[i]))
+		}
+	}
+
+	return columns
+}
+
 // ToTable takes an empty interfaces and returns a Table instance.
 func ToTable(arg interface{}) stmt.Table {
 	table := stmt.Table{}

--- a/builder/select.go
+++ b/builder/select.go
@@ -26,14 +26,14 @@ func (b Select) Distinct() Select {
 
 // Columns adds result columns to the query.
 func (b Select) Columns(args ...interface{}) Select {
-	if len(b.query.Columns) != 0 {
+	if len(b.query.Expressions) != 0 {
 		panic("loukoum: select builder has columns already defined")
 	}
 	if len(args) == 0 {
 		args = []interface{}{"*"}
 	}
 
-	b.query.Columns = ToColumns(args)
+	b.query.Expressions = ToSelectExpressions(args)
 
 	return b
 }

--- a/builder/select_test.go
+++ b/builder/select_test.go
@@ -9,7 +9,7 @@ import (
 	"github.com/ulule/loukoum/stmt"
 )
 
-func TestSelect(t *testing.T) {
+func TestSelect_Columns(t *testing.T) {
 	RunBuilderTests(t, []BuilderTest{
 		{
 			Name:      "Simple",
@@ -55,6 +55,71 @@ func TestSelect(t *testing.T) {
 				}),
 			},
 			SameQuery: "SELECT a, b, c",
+		},
+		{
+			Name:      "Exists expression",
+			Builder:   loukoum.Select(loukoum.Exists(loukoum.Select("1"))),
+			SameQuery: "SELECT EXISTS (SELECT 1)",
+		},
+		{
+			Name:      "Count expression",
+			Builder:   loukoum.Select(loukoum.Count("*")),
+			SameQuery: "SELECT COUNT(*)",
+		},
+		{
+			Name:      "Count expression with alias",
+			Builder:   loukoum.Select(loukoum.Count("*").As("counter")),
+			SameQuery: "SELECT COUNT(*) AS counter",
+		},
+		{
+			Name:      "Count expression with column",
+			Builder:   loukoum.Select(loukoum.Count("id")),
+			SameQuery: "SELECT COUNT(id)",
+		},
+		{
+			Name:      "Count expression with column and alias",
+			Builder:   loukoum.Select(loukoum.Count("id").As("counter")),
+			SameQuery: "SELECT COUNT(id) AS counter",
+		},
+		{
+			Name:      "Count expression with distinct and column",
+			Builder:   loukoum.Select(loukoum.Count("id").Distinct(true)),
+			SameQuery: "SELECT COUNT(DISTINCT id)",
+		},
+		{
+			Name:      "Count expression with distinct, column and alias",
+			Builder:   loukoum.Select(loukoum.Count("id").Distinct(true).As("counter")),
+			SameQuery: "SELECT COUNT(DISTINCT id) AS counter",
+		},
+		{
+			Name:      "Max expression with column",
+			Builder:   loukoum.Select(loukoum.Max("amount")),
+			SameQuery: "SELECT MAX(amount)",
+		},
+		{
+			Name:      "Max expression with column and alias",
+			Builder:   loukoum.Select(loukoum.Max("amount").As("max_amount")),
+			SameQuery: "SELECT MAX(amount) AS max_amount",
+		},
+		{
+			Name:      "Min expression with column",
+			Builder:   loukoum.Select(loukoum.Min("amount")),
+			SameQuery: "SELECT MIN(amount)",
+		},
+		{
+			Name:      "Min expression with column and alias",
+			Builder:   loukoum.Select(loukoum.Min("amount").As("min_amount")),
+			SameQuery: "SELECT MIN(amount) AS min_amount",
+		},
+		{
+			Name:      "Sum expression with column",
+			Builder:   loukoum.Select(loukoum.Sum("amount")),
+			SameQuery: "SELECT SUM(amount)",
+		},
+		{
+			Name:      "Sum expression with column and alias",
+			Builder:   loukoum.Select(loukoum.Sum("amount").As("sum_amount")),
+			SameQuery: "SELECT SUM(amount) AS sum_amount",
 		},
 	})
 }
@@ -901,12 +966,6 @@ func TestSelect_Exists(t *testing.T) {
 			),
 			Args: []interface{}{2},
 		},
-		// TODO (novln): Implement this use case.
-		// {
-		// 	Name:      "Column clause",
-		// 	Builder:   loukoum.Select(loukoum.Exists(loukoum.Select("1"))),
-		// 	SameQuery: "SELECT EXISTS (SELECT 1)",
-		// },
 	})
 }
 

--- a/loukoum.go
+++ b/loukoum.go
@@ -95,6 +95,26 @@ func Exists(value interface{}) stmt.Exists {
 	return stmt.NewExists(value)
 }
 
+// Count is a wrapper to create a new Count expression.
+func Count(value string) stmt.Count {
+	return stmt.NewCount(value)
+}
+
+// Max is a wrapper to create a new Max expression.
+func Max(value string) stmt.Max {
+	return stmt.NewMax(value)
+}
+
+// Min is a wrapper to create a new Min expression.
+func Min(value string) stmt.Min {
+	return stmt.NewMin(value)
+}
+
+// Sum is a wrapper to create a new Sum expression.
+func Sum(value string) stmt.Sum {
+	return stmt.NewSum(value)
+}
+
 // With is a wrapper to create a new WithQuery statement.
 func With(name string, value interface{}) stmt.WithQuery {
 	return stmt.NewWithQuery(name, value)

--- a/stmt/aggregate.go
+++ b/stmt/aggregate.go
@@ -1,0 +1,189 @@
+package stmt
+
+import (
+	"github.com/ulule/loukoum/token"
+	"github.com/ulule/loukoum/types"
+)
+
+// Count is a aggregate expression.
+type Count struct {
+	Value      Raw
+	IsDistinct bool
+	Alias      string
+}
+
+// NewCount returns a new Count instance.
+func NewCount(value string) Count {
+	return Count{
+		Value: NewRaw(value),
+	}
+}
+
+// As is used to give an alias name to the COUNT function.
+func (count Count) As(alias string) Count {
+	count.Alias = alias
+	return count
+}
+
+// Distinct is used to define if count has a distinct clause.
+func (count Count) Distinct(value bool) Count {
+	count.IsDistinct = value
+	return count
+}
+
+// Write exposes statement as a SQL query.
+func (count Count) Write(ctx types.Context) {
+	ctx.Write(token.Count.String())
+	ctx.Write("(")
+	if count.IsDistinct {
+		ctx.Write(token.Distinct.String())
+		ctx.Write(" ")
+	}
+	count.Value.Write(ctx)
+	ctx.Write(")")
+	if count.Alias != "" {
+		ctx.Write(" ")
+		ctx.Write(token.As.String())
+		ctx.Write(" ")
+		ctx.Write(count.Alias)
+	}
+}
+
+// IsEmpty returns true if statement is undefined.
+func (count Count) IsEmpty() bool {
+	return count.Value.IsEmpty()
+}
+
+func (Count) selectExpression() {}
+
+// Ensure that Count is an SelectExpression
+var _ SelectExpression = Count{}
+
+// Max is a aggregate expression.
+type Max struct {
+	Value Raw
+	Alias string
+}
+
+// NewMax returns a new Max instance.
+func NewMax(value string) Max {
+	return Max{
+		Value: NewRaw(value),
+	}
+}
+
+// As is used to give an alias name to the MAX function.
+func (max Max) As(alias string) Max {
+	max.Alias = alias
+	return max
+}
+
+// Write exposes statement as a SQL query.
+func (max Max) Write(ctx types.Context) {
+	ctx.Write(token.Max.String())
+	ctx.Write("(")
+	max.Value.Write(ctx)
+	ctx.Write(")")
+	if max.Alias != "" {
+		ctx.Write(" ")
+		ctx.Write(token.As.String())
+		ctx.Write(" ")
+		ctx.Write(max.Alias)
+	}
+}
+
+// IsEmpty returns true if statement is undefined.
+func (max Max) IsEmpty() bool {
+	return max.Value.IsEmpty()
+}
+
+func (Max) selectExpression() {}
+
+// Ensure that Max is an SelectExpression
+var _ SelectExpression = Max{}
+
+// Min is a aggregate expression.
+type Min struct {
+	Value Raw
+	Alias string
+}
+
+// NewMin returns a new Min instance.
+func NewMin(value string) Min {
+	return Min{
+		Value: NewRaw(value),
+	}
+}
+
+// As is used to give an alias name to the MIN function.
+func (min Min) As(alias string) Min {
+	min.Alias = alias
+	return min
+}
+
+// Write exposes statement as a SQL query.
+func (min Min) Write(ctx types.Context) {
+	ctx.Write(token.Min.String())
+	ctx.Write("(")
+	min.Value.Write(ctx)
+	ctx.Write(")")
+	if min.Alias != "" {
+		ctx.Write(" ")
+		ctx.Write(token.As.String())
+		ctx.Write(" ")
+		ctx.Write(min.Alias)
+	}
+}
+
+// IsEmpty returns true if statement is undefined.
+func (min Min) IsEmpty() bool {
+	return min.Value.IsEmpty()
+}
+
+func (Min) selectExpression() {}
+
+// Ensure that Min is an SelectExpression
+var _ SelectExpression = Min{}
+
+// Sum is a aggregate expression.
+type Sum struct {
+	Value Raw
+	Alias string
+}
+
+// NewSum returns a new Sum instance.
+func NewSum(value string) Sum {
+	return Sum{
+		Value: NewRaw(value),
+	}
+}
+
+// As is used to give an alias name to the SUM function.
+func (sum Sum) As(alias string) Sum {
+	sum.Alias = alias
+	return sum
+}
+
+// Write exposes statement as a SQL query.
+func (sum Sum) Write(ctx types.Context) {
+	ctx.Write(token.Sum.String())
+	ctx.Write("(")
+	sum.Value.Write(ctx)
+	ctx.Write(")")
+	if sum.Alias != "" {
+		ctx.Write(" ")
+		ctx.Write(token.As.String())
+		ctx.Write(" ")
+		ctx.Write(sum.Alias)
+	}
+}
+
+// IsEmpty returns true if statement is undefined.
+func (sum Sum) IsEmpty() bool {
+	return sum.Value.IsEmpty()
+}
+
+func (Sum) selectExpression() {}
+
+// Ensure that Sum is an SelectExpression
+var _ SelectExpression = Sum{}

--- a/stmt/column.go
+++ b/stmt/column.go
@@ -66,5 +66,7 @@ func (column Column) IsEmpty() bool {
 	return column.Name == ""
 }
 
-// Ensure that Column is a Statement
-var _ Statement = Column{}
+func (Column) selectExpression() {}
+
+// Ensure that Column is a SelectExpression.
+var _ SelectExpression = Column{}

--- a/stmt/expression.go
+++ b/stmt/expression.go
@@ -368,7 +368,7 @@ func (raw Raw) Write(ctx types.Context) {
 
 // IsEmpty returns true if statement is undefined.
 func (raw Raw) IsEmpty() bool {
-	return false
+	return raw.Value == ""
 }
 
 // Ensure that Raw is an Expression
@@ -387,6 +387,10 @@ type Wrapper struct {
 func NewWrapper(arg Expression) Expression {
 	switch value := arg.(type) {
 	case Select:
+		return &Wrapper{
+			Value: value,
+		}
+	case Exists:
 		return &Wrapper{
 			Value: value,
 		}

--- a/stmt/infix.go
+++ b/stmt/infix.go
@@ -47,13 +47,13 @@ func (expression InfixExpression) IsEmpty() bool {
 // And creates a new InfixExpression using given Expression.
 func (expression InfixExpression) And(value Expression) InfixExpression {
 	operator := NewAndOperator()
-	return NewInfixExpression(expression, operator, value)
+	return NewInfixExpression(expression, operator, NewWrapper(value))
 }
 
 // Or creates a new InfixExpression using given Expression.
 func (expression InfixExpression) Or(value Expression) InfixExpression {
 	operator := NewOrOperator()
-	return NewInfixExpression(expression, operator, value)
+	return NewInfixExpression(expression, operator, NewWrapper(value))
 }
 
 // Ensure that InfixExpression is an Expression

--- a/stmt/select.go
+++ b/stmt/select.go
@@ -5,21 +5,27 @@ import (
 	"github.com/ulule/loukoum/types"
 )
 
+// SelectExpression is a SQL expression for a SELECT statement.
+type SelectExpression interface {
+	Statement
+	selectExpression()
+}
+
 // Select is a SELECT statement.
 type Select struct {
-	Prefix   Prefix
-	With     With
-	Distinct bool
-	Columns  []Column
-	From     From
-	Joins    []Join
-	Where    Where
-	GroupBy  GroupBy
-	Having   Having
-	OrderBy  OrderBy
-	Limit    Limit
-	Offset   Offset
-	Suffix   Suffix
+	Prefix      Prefix
+	With        With
+	Distinct    bool
+	Expressions []SelectExpression
+	From        From
+	Joins       []Join
+	Where       Where
+	GroupBy     GroupBy
+	Having      Having
+	OrderBy     OrderBy
+	Limit       Limit
+	Offset      Offset
+	Suffix      Suffix
 }
 
 // NewSelect returns a new Select instance.
@@ -56,13 +62,13 @@ func (selekt Select) writeHead(ctx types.Context) {
 		ctx.Write(token.Distinct.String())
 	}
 
-	for i := range selekt.Columns {
+	for i := range selekt.Expressions {
 		if i == 0 {
 			ctx.Write(" ")
 		} else {
 			ctx.Write(", ")
 		}
-		selekt.Columns[i].Write(ctx)
+		selekt.Expressions[i].Write(ctx)
 	}
 
 	if !selekt.From.IsEmpty() {
@@ -117,7 +123,7 @@ func (selekt Select) writeTail(ctx types.Context) {
 
 // IsEmpty returns true if statement is undefined.
 func (selekt Select) IsEmpty() bool {
-	return len(selekt.Columns) == 0
+	return len(selekt.Expressions) == 0
 }
 
 func (Select) expression() {}

--- a/stmt/subquery.go
+++ b/stmt/subquery.go
@@ -1,18 +1,19 @@
 package stmt
 
 import (
+	"github.com/ulule/loukoum/token"
 	"github.com/ulule/loukoum/types"
 )
 
 // Exists is a subquery expression.
 type Exists struct {
-	subquery Expression
+	Subquery Expression
 }
 
 // NewExists returns a new Exists instance.
 func NewExists(value interface{}) Exists {
 	return Exists{
-		subquery: NewExpression(value),
+		Subquery: NewExpression(value),
 	}
 }
 
@@ -20,15 +21,18 @@ func (Exists) expression() {}
 
 // Write exposes statement as a SQL query.
 func (exists Exists) Write(ctx types.Context) {
-	ctx.Write("(EXISTS (")
-	exists.subquery.Write(ctx)
-	ctx.Write("))")
+	ctx.Write(token.Exists.String())
+	ctx.Write(" (")
+	exists.Subquery.Write(ctx)
+	ctx.Write(")")
 }
 
 // IsEmpty returns true if statement is undefined.
 func (exists Exists) IsEmpty() bool {
 	return false
 }
+
+func (Exists) selectExpression() {}
 
 // Ensure that Exists is an Expression
 var _ Expression = Exists{}

--- a/stmt/where.go
+++ b/stmt/where.go
@@ -13,7 +13,7 @@ type Where struct {
 // NewWhere returns a new Where instance.
 func NewWhere(expression Expression) Where {
 	return Where{
-		Condition: expression,
+		Condition: NewWrapper(expression),
 	}
 }
 
@@ -41,7 +41,7 @@ func (where Where) And(right Expression) Where {
 
 	left := where.Condition
 	operator := NewAndOperator()
-	where.Condition = NewInfixExpression(left, operator, right)
+	where.Condition = NewInfixExpression(left, operator, NewWrapper(right))
 	return where
 }
 
@@ -53,7 +53,7 @@ func (where Where) Or(right Expression) Where {
 
 	left := where.Condition
 	operator := NewOrOperator()
-	where.Condition = NewInfixExpression(left, operator, right)
+	where.Condition = NewInfixExpression(left, operator, NewWrapper(right))
 	return where
 }
 

--- a/token/token.go
+++ b/token/token.go
@@ -68,6 +68,11 @@ const (
 	Do        = Type("DO")
 	Nothing   = Type("NOTHING")
 	With      = Type("WITH")
+	Exists    = Type("EXISTS")
+	Count     = Type("COUNT")
+	Max       = Type("MAX")
+	Min       = Type("MIN")
+	Sum       = Type("SUM")
 )
 
 // A Token is defined by its type and a value.
@@ -113,6 +118,11 @@ var keywords = map[string]Type{
 	"DO":        Do,
 	"NOTHING":   Nothing,
 	"WITH":      With,
+	"EXISTS":    Exists,
+	"COUNT":     Count,
+	"MAX":       Max,
+	"MIN":       Min,
+	"SUM":       Sum,
 }
 
 // Lookup will try to map a statement to a keyword.

